src/cuda/api/array.hpp

```diff
@@ -0,0 +1,153 @@
+/**
+ * @file array.hpp
+ * 
+ * @brief A CUDA array wrapper class. Especially useful
+ * if used in combination with CUDA textures.
+ *
+ */
+
+#pragma once
+#ifndef CUDA_API_WRAPPERS_ARRAY_HPP
+#define CUDA_API_WRAPPERS_ARRAY_HPP
+
+#include <cuda/api/current_device.hpp>
+#include <cuda/api/error.hpp>
+
+#include <cuda_runtime.h>
+#include <cuda_runtime_api.h>
+
+#include <algorithm>
+#include <array>
+#include <numeric>
+
+namespace cuda {
+
+template<bool AssumedCurrent> class device_t;
+
+namespace array {
+
+namespace detail {
+
+/**
+ * @brief Do not use directly
+ *
+ */
+template <typename T, size_t NDIMS>
+class array_base {
+
+protected:
+	array_base(dimensions_t<NDIMS> dims, cudaArray* array_ptr) :
+	    dims_(dims),
+	    array_ptr_(array_ptr) {}
+
+public:
+	array_base()                        = delete;
+	array_base(const array_base& other) = delete;
+
+	array_base(array_base&& other) :
+	    array_ptr_(other.array_ptr_),
+	    dims_(other.dims_) {
+		other.array_ptr_ = nullptr;
+		other.dims_      = {0, 0, 0};
+	}
+
+	~array_base() {
+		if (array_ptr_) {
+			auto status = cudaFreeArray(array_ptr_);
+			throw_if_error(status, "failed freeing CUDA array");
```

Comment:
I agree. That should not happen. I mark them as `noexcept`. In case of an error the program terminates, which would be the behaviour according to 

https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rc-dtor-fail